can/can_graph.py

```diff
@@ -0,0 +1,294 @@
+        print(f'    {"source : " + component.name:20} : {str(component.description):85} : {str(component.help)}')
+
+def describe_plugins(plugins):
+    for plugin in plugins:
+        print(f'  {plugin.name:22} : {plugin.description} : {plugin.path}')
+
+        describe_components(plugin.source_component_classes.values())
+        describe_components(plugin.filter_component_classes.values())
+        describe_components(plugin.sink_component_classes.values())
+        print()
+
+
+# BT2 graph - dumps CANSource trace to stdout
+def graph_can_detail():
+    # Load required components from plugins
+    sink = system_plugins['text'].sink_component_classes['details']
+    source = user_plugins['can'].source_component_classes['CANSource']
+
+    # Create graph and add components
+    graph = bt2.Graph()
+
+    graph_sink = graph.add_component(sink, 'test_sink')
+    graph_source = graph.add_component(source, 'test_source',
+        # The plugin is actually capable of reading from multiple (list)
+        # input and database files
+        #
+        # For the sake of simplicity, this script only supplies ONE
+        # input and database file.
+        #
+        params=bt2.MapValue({
+            'inputs' : bt2.ArrayValue([CANSource_data_path]),
+            'databases' : bt2.ArrayValue([CANSource_dbc_path])
+        })
+    )
+
+    # Connect components together
+
+    # Do note that that input/output ports can be arbitrarily named
+    # The bt_plugin_can.CANSource uses the input file path for port name
+    #
+    # So we will ignore the port name and connect the first available port
+    # from the component
+    #
+    graph.connect_ports(
+        list(graph_source.output_ports.values())[0],
+        list(graph_sink.input_ports.values())[0]
+    )
+
+    # Run graph
+    graph.run()
+
+# BT2 graph - dumps CANSource trace to CTF
+def graph_can_ctf():
+    # Load required components from plugins
+    sink = system_plugins['ctf'].sink_component_classes['fs']
+    source = user_plugins['can'].source_component_classes['CANSource']
+
+    # Create graph and add components
+    graph = bt2.Graph()
+
+    graph_sink = graph.add_component(sink, 'test_sink',
+        params=bt2.MapValue({
+            'path': './ctf-full'
+        })
+    )
+
+    graph_source = graph.add_component(source, 'test_source',
+        params=bt2.MapValue({
+            'inputs': bt2.ArrayValue([CANSource_data_path]),
+            'databases': bt2.ArrayValue([CANSource_dbc_path])
+        })
+    )
+
+    # Connect components together
+    graph.connect_ports(
+        list(graph_source.output_ports.values())[0],
+        list(graph_sink.input_ports.values())[0]
+    )
+
+    # Run graph
+    graph.run()
+
+# BT2 graph - trims CANSource
+#
+# With the repo version of bt_plugin_gui, this will fail with
+#   Cannot make upstream message iterator initially seek
+#
+# You need to provide an implementation of
+#   def _user_seek_ns_from_origin(self, ns_from_origin):
```

Comment:
Having to add either `_user_seek_ns_from_origin` or `_user_seek_beginning` is expected for the trimmer to work (since in the end it needs to be able to `seek_ns_from_origin`).  But since you did add this method to `CANIterator`, I'm not sure it's really useful to mention this?